File: scripts/compute_rider_points.R
library(dplyr)
library(tidyr)
library(glue)
library(readr)

# ---- Parameters ----
source("config/config.R")

for (i in seq_len(nrow(EVENT_YEARS))) {
  id <- EVENT_YEARS$event_id[i]
  year <- EVENT_YEARS$event_year[i]

  # ---- Data inladen ----
  stages <- read.csv(glue("data/processed/stages/stages_{id}_{year}.csv"))
  results <- read.csv(glue("data/processed/results/{id}_{year}_all_stage_results.csv"))
  startlist <- read.csv(glue("data/processed/startlists_favorites/startlist_{id}_{year}.csv"))

  # ---- Stage results ----
  stage_results <- results %>%
    filter(category == "stage") %>%
    mutate(rank = as.integer(rank)) %>%
    filter(rank <= 10) %>%
    # Voeg een kolom toe die aangeeft of het een TTT is
    left_join(
      stages %>% select(stage_id, stage_name),
      by = "stage_id"
    ) %>%
    mutate(
      is_ttt = grepl("\\(TTT\\)", stage_name)
    ) %>%
    # kies het juiste puntenschema
    rowwise() %>%
    mutate(
      points = ifelse(is_ttt,
                      stage_points_ttt$points[match(rank, stage_points_ttt$rank)],
                      stage_points$points[match(rank, stage_points$rank)])
    ) %>%
    ungroup() %>%
    select(stage_id, rider_id, points)

  # ---- Daily class results ----
  daily_class_results <- results %>%
    filter(!grepl("stage-21", stage_id) & category %in% c("gc", "kom", "points", "youth")) %>%
    mutate(rank = as.integer(rank)) %>%
    filter(rank <= 3) %>%
    left_join(daily_class_points, by = c("category", "rank")) %>%
    select(stage_id, rider_id, points)

  # ---- Final class results ----
  final_class_results <- results %>%
    filter(grepl("stage-21", stage_id) & category %in% c("gc", "kom", "points", "youth")) %>%
    mutate(rank = as.integer(rank)) %>%
    filter((category == "gc" & rank <= 20) | rank <= 5) %>%
    left_join(final_class_points, by = c("category", "rank")) %>%
    mutate(stage_id = "final") %>%
    select(stage_id, rider_id, points)

  # ---- Alles samen ----
  all_points <- bind_rows(
    stage_results,
    daily_class_results,
    final_class_results
  ) %>%
    group_by(stage_id, rider_id) %>%
    summarise(stage_points = sum(points, na.rm = TRUE), .groups = "drop")

  # ---- Stage info ----
  stage_info <- stages %>%
    select(stage_id, stage = stage) %>%
    mutate(stage = as.integer(stage))

  final_stage_number <- max(stage_info$stage, na.rm = TRUE) + 1
  if ("final" %in% all_points$stage_id) {
    stage_info <- bind_rows(
      stage_info,
      tibble(stage_id = "final", stage = final_stage_number)
    )
  }

  # ---- Koppel renners (voor namen) ----
  all_points_named <- all_points %>%
    left_join(startlist %>% select(rider_id, rider_name), by = "rider_id") %>%
    left_join(stage_info, by = "stage_id")

  # ---- Cross join alle renners × beschikbare stages ----
  available_stage_ids <- unique(all_points_named$stage_id)

  all_combinations <- tidyr::crossing(
    stage_id = available_stage_ids,
    rider_id = unique(startlist$rider_id)
  ) %>%
    left_join(startlist %>% select(rider_id, rider_name), by = "rider_id") %>%
    left_join(stage_info, by = "stage_id")

  # ---- Join met berekende punten ----
  rider_stage_summary <- all_combinations %>%
    left_join(all_points_named %>% select(stage_id, rider_id, stage_points),
      by = c("stage_id", "rider_id")
    ) %>%
    mutate(stage_points = replace_na(stage_points, 0)) %>%
    arrange(rider_id, stage) %>%
    group_by(rider_id) %>%
    mutate(cumulative_points = cumsum(stage_points)) %>%
    ungroup() %>%
    select(rider_id, rider_name, stage_id, stage, stage_points, cumulative_points)

  # ---- Wegschrijven ----
  out_path <- glue("data/processed/points/rider_stage_summary_{id}_{year}.csv")
  write_csv(rider_stage_summary, out_path)

  message(glue("✅ Rider summary weggeschreven naar {out_path}"))
}

File: scripts/compute_rider_points.py
# scripts/compute_rider_points.py

import sys
import os
sys.path.insert(0, os.path.dirname(__file__))

import pandas as pd

from config import EVENT_YEARS, BASE_DIR, STAGE_POINTS, STAGE_POINTS_TTT, DAILY_CLASS_POINTS, FINAL_CLASS_POINTS


def main() -> None:
    for ev in EVENT_YEARS:
        event_id   = ev["event_id"]
        event_year = ev["event_year"]

        print(f"\n=== Computing rider points: {event_id} {event_year} ===")

        # ── Data inladen ──────────────────────────────────────────────────────
        results_file = BASE_DIR / "data" / "processed" / "results" / f"{event_id}_{event_year}_all_stage_results.csv"
        if not results_file.exists():
            print(f"  [SKIP] Geen resultaten beschikbaar voor {event_id} {event_year}")
            continue

        startlist = pd.read_csv(BASE_DIR / "data" / "processed" / "startlists_favorites" / f"startlist_{event_id}_{event_year}.csv")
        stages    = pd.read_csv(BASE_DIR / "data" / "processed" / "stages"               / f"stages_{event_id}_{event_year}.csv")
        results   = pd.read_csv(results_file)

        stage_info = stages[["stage_id", "stage"]].copy()
        stage_info["stage"] = pd.to_numeric(stage_info["stage"], errors="coerce").astype("Int64")

        last_stage_id = stage_info.loc[stage_info["stage"] == stage_info["stage"].max(), "stage_id"].iloc[0]
        ttt_stages    = set(stages.loc[stages["stage_name"].str.contains(r"\(TTT\)", na=False), "stage_id"])

        # ── Etappe punten ─────────────────────────────────────────────────────
        stage_results = (
            results[results["category"] == "stage"]
            .assign(rank=lambda d: pd.to_numeric(d["rank"], errors="coerce"))
            .dropna(subset=["rank"])
            .query("rank <= 10")
            .copy()
        )
        stage_results["is_ttt"] = stage_results["stage_id"].isin(ttt_stages)
        stage_results["points"] = stage_results.apply(
            lambda r: STAGE_POINTS_TTT.get(int(r["rank"]), 0) if r["is_ttt"]
                      else STAGE_POINTS.get(int(r["rank"]), 0),
            axis=1,
        )
        stage_results = stage_results[["stage_id", "rider_id", "points"]]

        # ── Dagelijkse klassementen ───────────────────────────────────────────
        daily_results = (
            results[
                (results["category"].isin(["gc", "kom", "points", "youth"])) &
                (results["stage_id"] != last_stage_id)
            ]
            .assign(rank=lambda d: pd.to_numeric(d["rank"], errors="coerce"))
            .dropna(subset=["rank"])
            .query("rank <= 3")
            .copy()
        )
        daily_results["points"] = daily_results.apply(
            lambda r: DAILY_CLASS_POINTS.get((r["category"], int(r["rank"])), 0), axis=1
        )
        daily_results = daily_results[["stage_id", "rider_id", "points"]]

        # ── Eindklassementen ──────────────────────────────────────────────────
        final_results = (
            results[
                (results["category"].isin(["gc", "kom", "points", "youth"])) &
                (results["stage_id"] == last_stage_id)
            ]
            .assign(rank=lambda d: pd.to_numeric(d["rank"], errors="coerce"))
            .dropna(subset=["rank"])
            .copy()
        )
        final_results = final_results[
            ((final_results["category"] == "gc") & (final_results["rank"] <= 20)) |
            ((final_results["category"] != "gc") & (final_results["rank"] <= 5))
        ].copy()
        final_results["points"]   = final_results.apply(
            lambda r: FINAL_CLASS_POINTS.get((r["category"], int(r["rank"])), 0), axis=1
        )
        final_results["stage_id"] = "final"
        final_results = final_results[["stage_id", "rider_id", "points"]]

        # ── Samenvoegen & per etappe sommeren ─────────────────────────────────
        all_points = (
            pd.concat([stage_results, daily_results, final_results], ignore_index=True)
            .groupby(["stage_id", "rider_id"], as_index=False)["points"]
            .sum()
            .rename(columns={"points": "stage_points"})
        )

        # Stage lookup uitbreiden met "final"
        final_stage_nr = int(stage_info["stage"].max()) + 1
        if not final_results.empty:
            stage_info = pd.concat([
                stage_info,
                pd.DataFrame({"stage_id": ["final"], "stage": [final_stage_nr]})
            ], ignore_index=True)

        all_points_named = (
            all_points
            .merge(startlist[["rider_id", "rider_name"]].drop_duplicates(), on="rider_id", how="left")
            .merge(stage_info, on="stage_id", how="left")
        )

        available_stage_ids = all_points_named["stage_id"].unique()

        # Cross join alle renners × beschikbare etappes
        all_combinations = (
            pd.DataFrame({"stage_id": available_stage_ids})
            .merge(startlist[["rider_id", "rider_name"]].drop_duplicates(), how="cross")
            .merge(stage_info, on="stage_id", how="left")
        )

        rider_summary = (
            all_combinations
            .merge(all_points_named[["stage_id", "rider_id", "stage_points"]],
                   on=["stage_id", "rider_id"], how="left")
            .fillna({"stage_points": 0})
            .sort_values(["rider_id", "stage"])
        )
        rider_summary["cumulative_points"] = rider_summary.groupby("rider_id")["stage_points"].cumsum()
        rider_summary = rider_summary[["rider_id", "rider_name", "stage_id", "stage", "stage_points", "cumulative_points"]]

        # ── Wegschrijven ──────────────────────────────────────────────────────
        out_dir = BASE_DIR / "data" / "processed" / "points"
        out_dir.mkdir(parents=True, exist_ok=True)
        out_file = out_dir / f"rider_stage_summary_{event_id}_{event_year}.csv"
        rider_summary.to_csv(out_file, index=False)
        print(f"  Opgeslagen: {out_file} ({len(rider_summary)} rijen)")
        print(f"=== Klaar: {event_id} {event_year} ===")


if __name__ == "__main__":
    main()
